homeassistant/components/adaptive_lighting/switch.py

```diff
@@ -0,0 +1,723 @@
+        today = self._relevant_events(now)
+        (_, prev_ts), (next_event, next_ts) = today
+        h, x = (  # pylint: disable=invalid-name
+            (prev_ts, next_ts)
+            if next_event in (SUN_EVENT_SUNSET, SUN_EVENT_SUNRISE)
+            else (next_ts, prev_ts)
+        )
+        k = 1 if next_event in (SUN_EVENT_SUNSET, SUN_EVENT_NOON) else -1
+        percentage = (0 - k) * ((now_ts - h) / (h - x)) ** 2 + k
+        return percentage
+
+    def _is_sleep(self):
+        return (
+            self._sleep_entity is not None
+            and self.hass.states.get(self._sleep_entity).state in self._sleep_state
+        )
+
+    def _calc_color_temp_kelvin(self):
+        if self._is_sleep():
+            return self._sleep_color_temp
+        if self._percent > 0:
+            delta = self._max_color_temp - self._min_color_temp
+            return (delta * self._percent) + self._min_color_temp
+        return self._min_color_temp
+
+    def _calc_brightness(self) -> float:
+        if self._is_sleep():
+            return self._sleep_brightness
+        if self._percent > 0:
+            return self._max_brightness
+        delta_brightness = self._max_brightness - self._min_brightness
+        percent = 1 + self._percent
+        return (delta_brightness * percent) + self._min_brightness
+
+    def _is_disabled(self):
+        return (
+            self._disable_entity is not None
+            and self.hass.states.get(self._disable_entity).state in self._disable_state
+        )
+
+    async def _adapt_light(self, light, transition, colors_only=False):
+        service_data = {ATTR_ENTITY_ID: light}
+        features = self._supported_features(light)
+
+        if "transition" in features:
+            if transition is None:
+                transition = self._transition
+            service_data[ATTR_TRANSITION] = transition
+
+        if "brightness" in features and self._adapt_brightness and not colors_only:
+            service_data[ATTR_BRIGHTNESS_PCT] = self._brightness
+
+        if (
+            "color_temp" in features
+            and self._adapt_color_temp
+            and not (self._prefer_rgb_color and "color" in features)
+        ):
+            attributes = self.hass.states.get(light).attributes
+            min_mireds, max_mireds = attributes["min_mireds"], attributes["max_mireds"]
+            color_temp_mired = max(min(self._color_temp_mired, max_mireds), min_mireds)
+            service_data[ATTR_COLOR_TEMP] = color_temp_mired
+        elif "color" in features and self._adapt_rgb_color:
+            service_data[ATTR_RGB_COLOR] = self._rgb_color
+
+        _LOGGER.debug(
+            "%s: Scheduling 'light.turn_on' with the following 'service_data': %s",
+            self._name,
+            service_data,
+        )
+
+        return self.hass.services.async_call(
+            LIGHT_DOMAIN,
+            SERVICE_TURN_ON,
+            service_data,
+            context=Context(),
```

Comment:
Is this useful at all? Does it do anything?

The thing I am trying to accomplish is to connect the switch to `turn_on` events.

Reply:
If you have an event object you can pass `event.context` here

Reply:
If you are reacting to another service call, you can forward it's context via self._context of the entity. But otherwise I'm not sure it is useful. I suppose one could setup a context so it's visible in logs that the change was done by a specific integration.

Would be neat to see that a change was done by adaptive lighting, but I don't think the context object have that feature yet.

Reply:
I am now passing the `event.context` but I would prefer to see "Turned on by Adaptive Lighting: living room".